File: Data Structures & Algorithms/4sum/submission-0.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        returnable = []


        for i in range(len(nums)):
            temp_list = []
            if i > 0 and nums[i] == nums[i-1]:
                continue

            for j in range(i+1, len(nums)):
                if j > i+1 and nums[j] == nums[j-1]:
                    continue

                temp_list_2 = []
                start = j+1
                end = len(nums) - 1

                while start < end:
                    temp = nums[i] + nums[j] + nums[start] + nums[end]

                    if temp > target:
                        end -= 1
                    elif temp < target:
                        start += 1
                    else:
                        sums = [nums[i], nums[j], nums[start], nums[end]]
                        if sums not in temp_list_2:
                            temp_list_2.append(sums)
                        start += 1

                for value in temp_list_2:
                    temp_list.append(value)

            for vals in temp_list:
                returnable.append(vals)

        return returnable

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    void find_combination(int index, vector<int>& candidates, int target, vector<int>& temp, vector<vector<int>>& returnable){
        if (index == candidates.size()) return;
        if (target < 0) return;
        if (target == 0){
            returnable.push_back(temp);
            return;
        }

        temp.push_back(candidates[index]);
        find_combination(index, candidates, target - candidates[index], temp, returnable);

        temp.pop_back();
        find_combination(index + 1, candidates, target, temp, returnable);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> returnable;
        vector<int> temp;

        find_combination(0, candidates, target, temp, returnable);
        return returnable;
    }
};